### PR TITLE
fix(node/syncer): Stop fetching header when all peers disconnected

### PR DESCRIPTION
**Pull request by oblique**
* Implemented `connecting_event_loop` and `connected_event_loop`
* In the `connecting_event_loop` Syncer waits for a trusted peer to connect and fetches the HEAD, then continues with the `connected_event_loop`.
* In the `connected_event_loop` Syncer starts the syncing process. If all peers disconnected, it moves back to the `connecting_event_loop`
* Added exponensial backoff for retrying the `try_init`
* Added a way to cancel spawned tasks in Syncer.

### CODE REVIEW

File: node/src/syncer.rs [hunk 1]
```diff
@@ -209,21 +304,33 @@ where
...
-                        Ok(height) => {
-                            tx.maybe_send(height);
-                            return;
-                        }
-                        Err(e) => {
-                            warn!("Intialization of subjective head failed: {e}. Trying again.")
-                        }
+        let fut = async move {
+            let mut backoff = ExponentialBackoffBuilder::default()
+                .with_max_interval(Duration::from_secs(60))
```
Comment (fl0rek) [inline node/src/syncer.rs:309]: Could you use const here?

**Review by fl0rek — COMMENTED** [review]
nice! left few nitpicks :)

File: node/src/executor.rs [hunk 2]
```diff
@@ -13,6 +15,19 @@ impl swarm::Executor for Executor {
     }
 }
 
+/// Spawn a task and that can be stopped by the `cancelation_token`.
+pub(crate) fn spawn_cancellable<F>(cancelation_token: CancellationToken, future: F)
```
[the remark this answers is not in the document]
Comment (oblique) [reply node/src/executor.rs:23]: Yes I was waiting to merge your PR

File: node/src/syncer.rs [hunk 3]
```diff
@@ -209,21 +304,33 @@ where
...
+                        tx.maybe_send(network_height);
+                        break;
+                    }
+                    Err(e) => {
+                        let sleep_dur = backoff
+                            .next_backoff()
+                            .expect("backoff never stops retrying");
+
+                        warn!("Intialization of subjective head failed: {e}. Trying again in {sleep_dur:?}.");
+                        tokio::time::sleep(sleep_dur).await;
```
[the remark this answers is not in the document]
Comment (oblique) [reply node/src/syncer.rs]: Yes I was waiting to merge your PR

File: node/src/syncer.rs [hunk 4]
```diff
@@ -151,19 +172,86 @@ where
...
-                    info!("Setting initial subjective head to {network_head_height}");
-                    self.subjective_head_height = Some(network_head_height);
+                    match self.subjective_head_height {
+                        Some(height) if network_head_height > height => {
+                            info!("Updating subjective head to {network_head_height}");
+                            self.subjective_head_height = Some(network_head_height);
+                        }
+                        Some(height) if network_head_height == height => {}
+                        Some(height) => {
+                            warn!("Network head ({network_head_height}) is smaller than the subjective head ({height})");
```
[the remark this answers is not in the document]
Comment (oblique) [reply node/src/syncer.rs]: I removed the whole behind of time handling, I realized that it creates much more edge cases and it doesn't worth the complexity.

File: node/src/syncer.rs (hunk 1 above)
Comment (oblique) [reply node/src/syncer.rs]: Done

File: node/src/syncer.rs [hunk 5]
```diff
@@ -79,22 +84,25 @@ where
...
             _store: PhantomData,
         })
     }
 
-    pub async fn stop(&self) -> Result<()> {
-        // TODO
-        Ok(())
+    pub fn stop(&self) {
+        // Singal the Worker to stop.
+        // TODO: Should we wait for the Worker to stop?
```
Comment (fl0rek) [inline node/src/syncer.rs:104]: I think that's a good idea in general, not necesarily in this PR, if it's a hassle.

File: node/src/syncer.rs [hunk 6]
```diff
@@ -303,28 +412,42 @@ where
             return;
         }
 
+        if self.p2p.peer_tracker_info().num_connected_peers == 0 {
+            // No connected peers. We can do the request.
```
Comment (fl0rek) [inline node/src/syncer.rs:416]: did you mean "can't" here?

**Review by fl0rek — APPROVED** [review]
nice 👍

File: node/src/syncer.rs (hunk 5 above)
Comment (oblique) [reply node/src/syncer.rs:104]: it is a bit, that is why i have it as todo

File: node/src/syncer.rs (hunk 6 above)
Comment (oblique) [reply node/src/syncer.rs]: fixed

**Pull request merged**